### PR TITLE
Refactor `Learner` and related configs to be more flexible and easier to use in a notebook

### PR DESCRIPTION
**Pull request by AdeelH**
## Overview

This PR makes the `Learner` more flexible and easier to use in a notebook. It also makes it possible to build datasets, model, loss, optimizer, and schedulers from configs directly outside of a `Learner`. This is particularly useful for examining/debugging datasets before getting started with training.

Major changes:
- Moves the dataset, model, loss, optimizer, schedulers creation code from `Learner` to the respective configs' `.build()` methods.
  - The `Learner`'s `build_*` methods are still there but now they just call `.build()` on the relevant config.
- It is now possible to pass any of or all of datasets, model, loss, optimizer, schedulers to the `Learner`'s constructor, which causes the `Learner` to skip building those specific objects.
- Since it is annoying to have to manually call `.update()` and `.validate_config()` on every config, these have been removed and the code has been moved to pydantic validators that get called automatically on initialization. See commit body for more details.
  - This refactoring has currently only been applied to configs in `rastervision_pytorch_learner`, but might make sense to apply to the rest too.
- `SolverConfig.ignore_last_class` has been replaced with the more general and consistent-with-pytorch `SolverConfig.ignore_class_index`.

Minor changes:
- Added a `show` flag to `Learner.plot_batch()` and `Learner.plot_predictions()` that causes the plot to be rendered to the frontend in addition to being saved as a file.
- Replaced `click` progress bars with `tqdm` progress bars.
- `LearnerConfig.model` can now be `None`.
- Updated unit tests.

### Checklist

- [x] Added `needs-backport` label if PR is bug fix that applies to previous minor release
- [x] Ran scripts/format_code and committed any changes
- [x] Documentation updated if needed
- [x] PR has a name that won't get you publicly shamed for vagueness

## Testing Instructions

See examples in this Colab notebook: https://colab.research.google.com/drive/1ybXLCNlbGRM9vAjI-FPj0S_lvz2wh_UR?usp=sharing

Connects to #1395

### CODE REVIEW

**Review by lewfish — APPROVED** [review]
I didn't review every line of code due to the size of the PR, but the high-level changes smell right to me.
